domain/relation/service/relation.go

```diff
@@ -333,6 +345,35 @@ func (s *LeadershipService) SetRelationApplicationSettings(
 	})
 }
 
+// SetRelationApplicationAndUnitSettings records settings for a unit and
+// an application in a relation.
+//
+// The following error types can be expected to be returned:
+//   - [corelease.ErrNotHeld] if the unit is not the leader.
+//   - [relationerrors.RelationUnitNotFound] is returned if the
+//     relation unit is not found.
+func (s *LeadershipService) SetRelationApplicationAndUnitSettings(
+	ctx context.Context,
+	unitName unit.Name,
+	relationUnitUUID corerelation.UnitUUID,
+	applicationSettings, unitSettings map[string]string,
+) error {
+	if len(applicationSettings) == 0 && len(unitSettings) == 0 {
+		return nil
+	}
+
+	if err := unitName.Validate(); err != nil {
+		return errors.Capture(err)
+	}
+	if err := relationUnitUUID.Validate(); err != nil {
+		return errors.Errorf(
+			"%w:%w", relationerrors.RelationUUIDNotValid, err)
+	}
+	return s.leaderEnsurer.WithLeader(ctx, unitName.Application(), unitName.String(), func(ctx context.Context) error {
+		return s.st.SetRelationApplicationAndUnitSettings(ctx, relationUnitUUID, applicationSettings, unitSettings)
```

Comment:
issue: a unit can set its own settings without being the leader.  This call is used for CommitHookChanges in the uniter facade, thus every unit will use it.

Reply:
This is a bit of an issue, our current architecture doesn't allow for a leadership check mid transaction (for good reason), but this is what was done in mongo. The `updateApplicationSettingsOp` checks for leadership and then returns the operations.

It seems that because the none-leader unit never tried to update the application settings it was never an issue that the transaction would fail.

It seems the best solution for us is to only check leadership if there are application settings getting changed. I will implement that.